ql/src/java/org/apache/hadoop/hive/ql/exec/tez/CustomPartitionVertex.java

```diff
@@ -533,83 +536,87 @@ private FileSplit getFileSplitFromEvent(InputDataInformationEvent event) throws
+        "CustomPartitionVertex.inputToBucketMap is not defined for " + inputName);
+    int inputBucketSize = isMainWork ? numBuckets : inputToBucketMap.get(inputName);
 
-    Multimap<Integer, InputSplit> bucketToInitialSplitMap =
-        ArrayListMultimap.create();
+    Multimap<Integer, InputSplit> bucketToSplitMap = ArrayListMultimap.create();
 
     boolean fallback = false;
-    Map<Integer, Integer> bucketIds = new HashMap<>();
     for (Map.Entry<String, Set<FileSplit>> entry : pathFileSplitsMap.entrySet()) {
       // Extract the buckedID from pathFilesMap, this is more accurate method,
       // however. it may not work in certain cases where buckets are named
       // after files used while loading data. In such case, fallback to old
       // potential inaccurate method.
       // The accepted file names are such as 000000_0, 000001_0_copy_1.
-      String bucketIdStr =
-              Utilities.getBucketFileNameFromPathSubString(entry.getKey());
+      String bucketIdStr = Utilities.getBucketFileNameFromPathSubString(entry.getKey());
       int bucketId = Utilities.getBucketIdFromFile(bucketIdStr);
       if (bucketId == -1) {
         fallback = true;
-        LOG.info("Fallback to using older sort based logic to assign " +
-                "buckets to splits.");
-        bucketIds.clear();
+        LOG.info("Fallback to using older sort based logic to assign buckets to splits.");
+        bucketToSplitMap.clear();
         break;
       }
-      // Make sure the bucketId is at max the numBuckets
-      bucketId = bucketId % numBuckets;
-      bucketIds.put(bucketId, bucketId);
+
+      // Utilities.getBucketIdFromFile() returns negative value only if it fails to retrieve bucketID.
+      Preconditions.checkState(bucketId >= 0);
+
+      if (bucketId >= inputBucketSize) {
+        int newBucketId = bucketId % inputBucketSize;
+        LOG.info("The bucketID" + bucketId + " for file " + entry.getKey() + " is not acceptable. " +
+            "The bucket size of input " + inputName + " is " + inputBucketSize + ". " +
+            "Use " + newBucketId + " instead.");
+        bucketId = newBucketId;
+      }
+
       for (FileSplit fsplit : entry.getValue()) {
-        bucketToInitialSplitMap.put(bucketId, fsplit);
+        bucketToSplitMap.put(bucketId, fsplit);
       }
     }
 
-    int bucketNum = 0;
     if (fallback) {
       // This is the old logic which assumes that the filenames are sorted in
       // alphanumeric order and mapped to appropriate bucket number.
+      int bucketNum = 0;
       for (Map.Entry<String, Set<FileSplit>> entry : pathFileSplitsMap.entrySet()) {
-        int bucketId = bucketNum % numBuckets;
         for (FileSplit fsplit : entry.getValue()) {
-          bucketToInitialSplitMap.put(bucketId, fsplit);
+          bucketToSplitMap.put(bucketNum, fsplit);
+        }
+        bucketNum = bucketNum + 1;
+
+        if (bucketNum == inputBucketSize) {
```

Comment:
That was my mistake; bucketNum should be reset to 0 if it reaches the inputBucketSize. Thank you for pointing that out. I fixed this code to use modulo operator instead of if clause, which is similar to current implementation.